### PR TITLE
CORDA-1524 Fix NodeSchedulerService

### PR DESCRIPTION
**Pull request by blsemo**
The NodeSchedulerService used to have a race condition where it could happen that two flows for the same state get started, leading to an attempted double spend.
This got cleared up by giving the scheduler an event loop on a worker thread, and doing all selection of states etc. on this single thread. The external calls now only trigger this thread to recheck what to execute rather than getting involved in any of the selection logic.

### CODE REVIEW

File: .idea/compiler.xml [hunk 1]
```diff
@@ -71,6 +71,8 @@
       <module name="example-code_integrationTest" target="1.8" />
       <module name="example-code_main" target="1.8" />
       <module name="example-code_test" target="1.8" />
+      <module name="experimental-blobinspector_main" target="1.8" />
```
Comment (rick-r3) [inline .idea/compiler.xml]: Don't think this should be changed.

Comment (blsemo) [reply .idea/compiler.xml]: Oh well, the joy of checking in `.idea/compiler.xml`

File: node/src/main/kotlin/net/corda/node/services/events/NodeSchedulerService.kt [hunk 2]
```diff
@@ -141,26 +142,32 @@ class NodeSchedulerService(private val clock: CordaClock,
...
+    // any new state should trigger a reschedule immediately if nothing is scheduled, so I would not expect
+    // this to usually trigger anything.
+    private val idleWaitSeconds = 60.seconds
+
     // We need the [StateMachineManager] to be constructed before this is called in case it schedules a flow.
     fun start() {
-        mutex.locked {
-            rescheduleWakeUp()
-        }
+        schedulerTimerExecutor.execute{runLoopFunction()}
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/events/NodeSchedulerService.kt]: whitespace

File: node/src/main/kotlin/net/corda/node/services/events/NodeSchedulerService.kt [hunk 3]
```diff
@@ -141,26 +142,32 @@ class NodeSchedulerService(private val clock: CordaClock,
...
     }
 
     override fun scheduleStateActivity(action: ScheduledStateRef) {
         log.trace { "Schedule $action" }
         // Only increase the number of unfinished schedules if the state didn't already exist on the queue
-        val countUp = !schedulerRepo.merge(action)
+        if( !schedulerRepo.merge(action) ){
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/events/NodeSchedulerService.kt]: whitespace

File: node/src/main/kotlin/net/corda/node/services/events/NodeSchedulerService.kt [hunk 4]
```diff
@@ -186,42 +193,45 @@ class NodeSchedulerService(private val clock: CordaClock,
...
+            if (scheduledState != null) {
                 if (!awaitWithDeadline(clock, scheduledState.scheduledAt, ourRescheduledFuture)) {
                     log.trace { "Invoking as next $scheduledState" }
                     onTimeReached(scheduledState)
                 } else {
                     log.trace { "Rescheduled $scheduledState" }
                 }
             }
+            else {
+                awaitWithDeadline(clock, Instant.now() + idleWaitSeconds , ourRescheduledFuture )
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/events/NodeSchedulerService.kt]: Technically, `Instant.now()` should be `Clock.instant()` to work with test clocks.

File: node/src/main/kotlin/net/corda/node/services/events/NodeSchedulerService.kt [hunk 5]
```diff
@@ -186,42 +193,45 @@ class NodeSchedulerService(private val clock: CordaClock,
...
-            nextScheduledAction = schedulerRepo.getLatest(startingStateRefs.size + 1).firstOrNull { !startingStateRefs.contains(it.second) }?.second
-            Pair(nextScheduledAction, rescheduled!!)
-        }
-        if (scheduledState != null) {
-            schedulerTimerExecutor.execute {
-                log.trace(schedulingAsNextFormat, scheduledState)
-                // This will block the scheduler single thread until the scheduled time (returns false) OR
-                // the Future is cancelled due to rescheduling (returns true).
+    private fun runLoopFunction(){
+        while (mutex.locked { running }){
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/events/NodeSchedulerService.kt]: We can eliminate the mutex locking here by making `running` `@Volatile`

Comment (blsemo) [reply node/src/main/kotlin/net/corda/node/services/events/NodeSchedulerService.kt]: I quite like the mutex at this point - we need it anyway at the place where the flag is set (as it needs to be locked when we trigger a reschedule to pick up the shutdown), so locking it here makes sure we don't enter the loop while a shutdown is being triggered.

Comment (rick-r3) [reply node/src/main/kotlin/net/corda/node/services/events/NodeSchedulerService.kt]: Okay.  Not too fussed.

File: node/src/test/kotlin/net/corda/node/services/events/NodeSchedulerServiceTest.kt [hunk 6]
```diff
@@ -200,7 +232,6 @@ class NodeSchedulerServiceTest : NodeSchedulerServiceTestBase() {
     fun `test activity due in the future and schedule another for same time`() {
         val eventA = schedule(mark + 1.days)
         val eventB = schedule(mark + 1.days)
-        assertWaitingFor(eventA)
```
Comment (rick-r3) [inline node/src/test/kotlin/net/corda/node/services/events/NodeSchedulerServiceTest.kt]: Whether we are waiting for A or B is an implementation detail that no longer holds.

File: node/src/test/kotlin/net/corda/node/services/events/NodeSchedulerServiceTest.kt [hunk 7]
```diff
@@ -88,7 +120,7 @@ open class NodeSchedulerServiceTestBase {
 }
 
 class MockScheduledFlowRepository : ScheduledFlowRepository {
-    private val map = HashMap<StateRef, ScheduledStateRef>()
+    private val map = ConcurrentHashMap<StateRef, ScheduledStateRef>()
```
Comment (rick-r3) [inline node/src/test/kotlin/net/corda/node/services/events/NodeSchedulerServiceTest.kt]: Concurrent updates mean we need concurrent map.

File: node/src/test/kotlin/net/corda/node/services/events/NodeSchedulerServiceTest.kt [hunk 8]
```diff
@@ -68,16 +86,30 @@ open class NodeSchedulerServiceTestBase {
...
+        doAnswer {
+            traces += it.getArgument<ScheduledStateRef>(1)
+        }.whenever(it).trace(eq(NodeSchedulerService.schedulingAsNextFormat), any<Object>())
     }
 
-    protected fun assertWaitingFor(ssr: ScheduledStateRef, total: Int = 1) {
-        // The timeout is to make verify wait, which is necessary as we're racing the NSS thread i.e. we often get here just before the trace:
-        verify(log, timeout(5000).times(total)).trace(NodeSchedulerService.schedulingAsNextFormat, ssr)
+    protected fun assertWaitingFor(ssr: ScheduledStateRef) {
+        val endTime = System.currentTimeMillis() + TimeUnit.SECONDS.toMillis(5)
```
Comment (rick-r3) [inline node/src/test/kotlin/net/corda/node/services/events/NodeSchedulerServiceTest.kt]: This is slightly stricter and looks at what the last log entry is.  Also, the previous count of entries didn't make sense with the revised implementation of the scheduler.

**Pull request merged**